File: netlist.py
from dataclasses import dataclass


@dataclass(frozen=True)
class NetlistGate:
    gate_type: str
    name: str
    connections: list[str]


@dataclass(frozen=True)
class NetlistAssign:
    target: str
    source: str


@dataclass(frozen=True)
class NetlistModule:
    name: str
    ports: list[str]
    inputs: list[str]
    outputs: list[str]
    wires: list[str]
    gates: list[NetlistGate]
    assigns: list[NetlistAssign]


def emit_verilog(netlist: NetlistModule) -> str:
    lines: list[str] = []
    ports = ", ".join(netlist.ports)
    lines.append(f"module {netlist.name}({ports});")
    if netlist.inputs:
        lines.append(f"  input {', '.join(netlist.inputs)};")
    if netlist.outputs:
        lines.append(f"  output {', '.join(netlist.outputs)};")
    if netlist.wires:
        lines.append(f"  wire {', '.join(netlist.wires)};")
    for gate in netlist.gates:
        lines.append(f"  {gate.gate_type} {gate.name}({', '.join(gate.connections)});")
    for assign in netlist.assigns:
        lines.append(f"  assign {assign.target} = {assign.source};")
    lines.append("endmodule")
    return "\n".join(lines) + "\n"

File: src/netlist.ts
export type NetlistGate = {
  gateType: string;
  name: string;
  connections: string[];
};

export type NetlistAssign = {
  target: string;
  source: string;
};

export type NetlistModule = {
  name: string;
  ports: string[];
  inputs: string[];
  outputs: string[];
  wires: string[];
  gates: NetlistGate[];
  assigns: NetlistAssign[];
};

export function emitVerilog(netlist: NetlistModule): string {
  const lines: string[] = [];

  lines.push(`module ${netlist.name}(${netlist.ports.join(", ")});`);

  if (netlist.inputs.length > 0) {
    lines.push(`  input ${netlist.inputs.join(", ")};`);
  }

  if (netlist.outputs.length > 0) {
    lines.push(`  output ${netlist.outputs.join(", ")};`);
  }

  if (netlist.wires.length > 0) {
    lines.push(`  wire ${netlist.wires.join(", ")};`);
  }

  for (const gate of netlist.gates) {
    lines.push(`  ${gate.gateType} ${gate.name}(${gate.connections.join(", ")});`);
  }

  for (const assign of netlist.assigns) {
    lines.push(`  assign ${assign.target} = ${assign.source};`);
  }

  lines.push("endmodule");

  return `${lines.join("\n")}\n`;
}
